[IMP] estate: Linked business logic to action buttons to cancel, set as sold, accept, or refuse offers following Odoo Guide t#70762

Cindypsd opened the pull request:
Explanation:

Added buttons to cancel or set a property as sold, a canceled property cannot be sold and a sold property cannot be canceled, it should raise an error in any of these cases:

![Captura de Pantalla 2023-04-25 a la(s) 12 58 12](https://user-images.githubusercontent.com/108036643/234378088-badb9a3f-dae8-44c8-8c60-f15558608534.png)

Added a function  to accept or refuse an offer:
![Captura de Pantalla 2023-04-25 a la(s) 13 01 03](https://user-images.githubusercontent.com/108036643/234378370-1257feca-b689-417b-98df-99383a03dc5a.png)

Once an offer is accepted, the selling price and the buyer should be set:
![Captura de Pantalla 2023-04-25 a la(s) 13 12 32](https://user-images.githubusercontent.com/108036643/234379350-715fc0b0-b636-45b9-9ad2-9b4425fb5f94.png)


![Captura de Pantalla 2023-04-25 a la(s) 13 01 11](https://user-images.githubusercontent.com/108036643/234378637-fa0c485d-cbd7-42d1-bcec-b79852bcc411.png)

sebasdrk17 commented on estate/models/estate_property.py:
```diff
@@ -78,3 +78,24 @@ def _onchange_garden(self):
+
+    def action_sold(self):
+        for record in self:
+            if record.state == "canceled":
+                raise exceptions.UserError("Canceled property cannot be sold")
+            elif record.state == "sold":
+                raise exceptions.UserError("Property is already sold")
```
This is OK if we need specific error messages

another alternative could be this:

```py
if record.state in ["canceled", "sold"]:
    raise exceptions.UserError("A property with state %s cannot be changed to sold." % record.state)
```

sebasdrk17 commented on estate/models/estate_property_offer.py:
```diff
@@ -14,17 +14,54 @@ class PropertyOffer(models.Model):
         ],
         copy=False,
     )
+    buyer_id = fields.Many2one("res.partner")
+    buyer_id = fields.Many2one("res.partner")
```
no need for two fields, remove one please

sebasdrk17 commented on estate/models/estate_property_offer.py:
```diff
@@ -14,17 +14,54 @@ class PropertyOffer(models.Model):
+
+    def _inverse_date_deadline(self):
+        for record in self:
+            if record.date_deadline:
```
if there is no `date_deadline` this is not going to do anything, is this expected?

sebasdrk17 commented on estate/models/estate_property_offer.py:
```diff
@@ -14,17 +14,54 @@ class PropertyOffer(models.Model):
-            record.date_deadline = fields.Datetime.now() + timedelta(days=(record.validity))
+            record.date_deadline = fields.Date.today() + timedelta(days=(record.validity))
 
+    def action_confirm(self):
+        for record in self:
+            if record.property_id.state == "offer_accepted":
+                raise exceptions.UserError_("More than one offer cannot be accepted")
+            record.status = "accepted"
+            record.property_id.selling_price = record.price
+            record.property_id.state = "offer_accepted"
+            record.property_id.buyer_id = record.buyer_id
+        return True
+
+    def action_confirm(self):
+        for record in self:
+            if record.property_id.state == "offer_accepted":
+                raise exceptions.UserError_("More than one offer cannot be accepted")
+            record.status = "accepted"
+            record.property_id.selling_price = record.price
+            record.property_id.state = "offer_accepted"
+            record.property_id.buyer_id = record.buyer_id
+        return True
```
Duplicated method, remove one please

Cindypsd commented:
@sebasdrk17 

Requested changes made

sebasdrk17 reviewed (changes requested):
Functional testing:

This warning triggers when loading the instance:

![Screenshot 2023-04-26 at 10 41 45 a m](https://user-images.githubusercontent.com/47955967/234645001-50d70892-fb93-4e8f-b88c-c2467126234c.png)

I found something weird, but tell me if that's how it's supposed to be:

https://user-images.githubusercontent.com/47955967/234649721-eda7b484-da8e-4ac0-b246-da65ec9e6d5d.mp4

Cindypsd commented:
@sebasdrk17 

I have made the necessary changes to remove the warnings. Additionally, I updated the function action_sold to ensure that when an offer is accepted and the property is set as sold, the selling price matches the accepted offer.

![Captura de Pantalla 2023-04-26 a la(s) 12 20 21](https://user-images.githubusercontent.com/108036643/234667652-860a33b5-a076-4775-bde3-de69e5a06cb0.png)


Regarding the labels on the form view of the property, I didn't add them before because I was following the example in the guide. However, I have now added a placeholder for the title and a label for the tags. 

![Captura de Pantalla 2023-04-26 a la(s) 12 17 59](https://user-images.githubusercontent.com/108036643/234667822-5e39288a-9035-4f4d-ae85-95b98475eac2.png)

sebasdrk17 commented:
Please squash your commits, only one commit should be merged (oldest PR were merged without squash, my bad)

Cindypsd commented:
@sebasdrk17  Done, commits squashed
